Firestore/Example/Tests/GoogleTest/FSTGoogleTestTests.mm

```diff
@@ -0,0 +1,181 @@
+      if (strcmp(testInfoName, testInfo->name()) == 0) {
+        return testInfo;
+      }
+    }
+  }
+
+  XCTFail(@"Failed to find a TestInfo corresponding to class %@ method %@", testClassName,
+          selectorName);
+  return nullptr;
+}
+
+/** Finds the TestInfo corresponding to this test and reports it to XCTest */
+void ReportTestResult(XCTestCase *self, SEL _cmd) {
+  const testing::TestInfo *testInfo = FindTestInfo(self, _cmd);
+  if (!testInfo) {
+    return;
+  }
+
+  if (!testInfo->should_run()) {
+    // Test was filtered out by gunit; nothing to report.
+    return;
+  }
+
+  const testing::TestResult *result = testInfo->result();
+  if (result->Passed()) {
+    // Let XCode know that the test ran and succeeded.
+    XCTAssertTrue(true);
+    return;
+  }
+
+  // Test failed :-(. Record the failure such that XCode will navigate directly to the file:line.
+  int parts = result->total_part_count();
+  for (int i = 0; i < parts; i++) {
+    const testing::TestPartResult &part = result->GetTestPartResult(i);
+
+    NSString *message = [NSString stringWithCString:part.message() encoding:NSUTF8StringEncoding];
+    NSString *fileName = @"";
+    if (part.file_name()) {
+      fileName = [NSString stringWithCString:part.file_name() encoding:NSUTF8StringEncoding];
+    }
+    NSUInteger lineNumber = 0;
+    if (part.line_number() > 0) {
+      lineNumber = part.line_number();
+    }
+
+    [self recordFailureWithDescription:message inFile:fileName atLine:lineNumber expected:YES];
+  }
+}
+
+/**
+ * Generates a new subclass of XCTestCase for the given GoogleTest TestCase. Each TestInfo (which
+ * represents an indivudal test method execution) is translated into a method on the test case.
+ *
+ * @return A new Class that's a subclass of XCTestCase, that's been registered with the Objective-C
+ *     runtime.
+ */
+Class CreateXCTestCaseClass(const testing::TestCase *testCase) {
+  NSString *testCaseName = [NSString stringWithFormat:@"%sTests", testCase->name()];
+  Class testClass = objc_allocateClassPair([XCTestCase class], [testCaseName UTF8String], 0);
+
+  int testInfos = testCase->total_test_count();
+  for (int j = 0; j < testInfos; j++) {
+    const testing::TestInfo *testInfo = testCase->GetTestInfo(j);
+
+    NSString *selectorName = [NSString stringWithFormat:@"test%s", testInfo->name()];
+    SEL selector = sel_registerName([selectorName UTF8String]);
+    IMP method = reinterpret_cast<IMP>(ReportTestResult);
+    class_addMethod(testClass, selector, method, "v@:");
+  }
+
+  objc_registerClassPair(testClass);
+  return testClass;
+}
+
+}  // namespace
+
+@implementation FSTGoogleTestTests
+
++ (XCTestSuite *)defaultTestSuite {
+  // XCTest calls +defaultTestSuite on any class that extends XCTestCase to generate the suite
+  // for that class. Override that to contain a new parent suite containing the default test suite
+  // containing test-methods in this class and an additional suite for each GoogleTest TestCase.
```

Comment:
I don't understand this comment very well. s/Override that/We override that/ would help, but I'm also fuzzy on this "parent suite" and "default test suite" and "additional suite" stuff. I don't care very much, since this is probably write-once-change-never code, but if you feel so compelled, I think there's probably room to make these comments a bit clearer.

Reply:
I've completely reworked this to mostly not use this structure but this comment is now gone.